### PR TITLE
challenger: init at 0.10.0

### PR DESCRIPTION
**Pull request by wegank**
## Description of changes

https://git.taler.net/challenger.git
https://taler.net/en/news/2024-06.html

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/ch/challenger/package.nix [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+  src = fetchgit {
+    url = "https://git.taler.net/challenger.git";
+    rev = "v${finalAttrs.version}";
+    hash = "sha256-fjT3igPQ9dQtOezwZVfK5fBaL22FKOCbjUF0U1urK0g=";
+  };
+
+  # https://git.taler.net/challenger.git/tree/bootstrap
+  preAutoreconf = ''
+    cd contrib
```
Comment (AndersonTorres) [inline pkgs/by-name/ch/challenger/package.nix:34]:
```suggestion
    pushd contrib
```

File: pkgs/by-name/ch/challenger/package.nix [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    rm -f Makefile.am
+    find wallet-core/challenger/ -type f -printf '  %p \\\n' | sort > Makefile.am.ext
+    # Remove extra '\' at the end of the file
+    truncate -s -2 Makefile.am.ext
+    cat Makefile.am.in Makefile.am.ext >> Makefile.am
+    # Prevent accidental editing of the generated Makefile.am
+    chmod -w Makefile.am
+    cd ..
```
Comment (AndersonTorres) [inline pkgs/by-name/ch/challenger/package.nix:42]:
```suggestion
    popd
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/by-name/ch/challenger/package.nix (hunk 1 above)
Comment (drupol) [reply pkgs/by-name/ch/challenger/package.nix]: In this context, what's the difference between using `cd` or `pushd` ?

Comment (wegank) [reply pkgs/by-name/ch/challenger/package.nix]: pushd/popd is probably less error-prone...

Comment (drupol) [reply pkgs/by-name/ch/challenger/package.nix]: I have no clue, I'm asking @AndersonTorres actually :)

File: pkgs/by-name/ch/challenger/package.nix [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    libtool
+    postgresql
+    taler-exchange
+    taler-merchant
+  ];
+
+  preFixup = ''
+    substituteInPlace $out/bin/challenger-{dbconfig,send-post.sh} \
+      --replace-fail "/bin/bash" "${runtimeShell}"
+  '';
```
Comment (drupol) [inline pkgs/by-name/ch/challenger/package.nix:72]: Isn't this automatically done?

Comment (wegank) [reply pkgs/by-name/ch/challenger/package.nix:72]: No.

Comment (drupol) [reply pkgs/by-name/ch/challenger/package.nix:72]: Isn't the `fixup` phase is not changing shebang of `.sh` files by default?

Comment (wegank) [reply pkgs/by-name/ch/challenger/package.nix:72]: `patchShebangs --host` will do nothing if there are no substitutes for free or in `buildInputs`.

Comment (drupol) [reply pkgs/by-name/ch/challenger/package.nix:72]: OK TIL, thanks!

File: pkgs/by-name/ch/challenger/package.nix [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+  meta = {
```
Comment (drupol) [inline pkgs/by-name/ch/challenger/package.nix:74]: No `meta.mainProgram` ?

Comment (wegank) [reply pkgs/by-name/ch/challenger/package.nix:74]: Taler components usually come with a bunch of binaries for different purposes, so there's probably no main program. The same goes for `taler-exchange`, `taler-merchant`, or even `gnunet`.

Comment (drupol) [reply pkgs/by-name/ch/challenger/package.nix:74]: Fair enough!

**Comment by drupol** [comment]
Thank you!

**Pull request merged**